File: frontend/src/store/models/Product.js
/** @module store/models/product */

import { EntityBase, EntityCollectionBase } from "./Entity";

/**
 * @class
 * @classdesc Represents a collection of products.
 */
export class ProductCollection extends EntityCollectionBase {
  constructor() {
    super(Product, "/api/v1/products")
  }
}

/**
 * @class
 * @classdesc Represents a product.
 */
export class Product extends EntityBase {
  constructor() {
    super();
    this.name = null;
    this.description = null;
    this.price = null;
    this.latitude = null;
    this.longitude = null;
  }

  /**
   * Updates the product from a data object with mappable key value pairs.
   * @param {object} data
   * @return {this}
   */
  update({name,description,price,latitude,longitude}) {
    this.name = name ? name : this.name;
    this.description = description ? description : this.description;
    this.price = price ? price : this.price;
    this.latitude = latitude ? latitude : this.latitude;
    this.longitude = longitude ? longitude : this.longitude;

    return this;
  }
}

File: backend/models/Product.py
from .mixins.Entity import EntityMixin
from .db import db


class Product(db.Model, EntityMixin):
    __tablename__ = "products"

    name = db.Column(
        db.String(128),
        nullable=False)
    description = db.Column(
        db.String(),
        nullable=True)
    price = db.Column(
        db.Float(precision=2),
        nullable=False,
        default=00.00)
    latitude = db.Column(
        db.Float(precision=5),
        nullable=False,
        default=00.00000)
    longitude = db.Column(
        db.Float(precision=5),
        nullable=False,
        default=000.00000)

    def to_dict(self):
        return {
            **self.entity_to_dict(),
            "name": self.name,
            "description": self.description,
            "price": self.price,
            "latitude": self.latitude,
            "longitude": self.longitude,
        }

    def __init__(self, name, description, price, latitude, longitude):
        self.name = name
        self.description = description
        self.price = price
        self.latitude = latitude
        self.longitude = longitude
